_posts/2018-11-11-fs.md

```diff
@@ -0,0 +1,81 @@
+title: "The choices we make: Impact of using host filesystem interface for secure containers"
+---
+
+The common theme behind all the secure container runtimes--Kata, Gvisor and Nabla--is to improve container isolation.
+However, to achieve the same goal, each solution adopts a different approach. 
+Kata employs virtual machine isolation, while Gvisor and Nabla use libOS+Seccomp based attack surface minimization.
+
+In this blog post, we highlight the impact of a key design choice difference between Nabla and Kata containers--the *touch points* into the host kernel from a filesystem perspective.
+
+Nabla uses a LibOS-powered filesystem interface, whereas Kata directly uses the [host's filesystem interface](https://github.com/kata-containers/documentation/blob/master/architecture.md#storage), exposed via the [9p](https://www.kernel.org/doc/Documentation/filesystems/9p.txt) protocol<sup>1</sup>.
+This means that Kata has more kernel *touch points* (filesystem and block device components), as compared to Nabla (block device only).
+And thus, in case of the former, there is a greater probability of hitting a kernel bug, or exploiting a vulnerability.
+
+
+### Kernel Oops!
+We highlight this by a recently discovered bug in the ext4 filesystem code- [CVE-2018-10840](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2018-10840).
+In this case, removing the [extended attributes](http://man7.org/linux/man-pages/man5/attr.5.html) of a file on a crafted ext4 image, leads to a kernel oops due to a buffer overflow.
+The culprit syscall which allows a userspace program to trigger this bug is [`removexattr()`](http://man7.org/linux/man-pages/man2/removexattr.2.html).
+When testing with a Ubuntu 18.04<sup>2</sup> host system, we were able to trigger this bug in the kernel, from inside a Kata container.
+The following video shows a demo of this.
+
+[TODO: video]
+
+
+### The Bug
+We used the [original author's](https://bugzilla.kernel.org/show_bug.cgi?id=199347) crafted image to test this bug, which can be exposed with the following simple trigger code:
+
+```
+// mpoint is the path to the mounted crafted image
+static void activity(char *mpoint) {
+  char *xattr;
+  int err = asprintf(&xattr, "%s/foo/bar/xattr", mpoint);
+  char buf2[113];
+  memset(buf2, 0, sizeof(buf2));
+  listxattr(xattr, buf2, sizeof(buf2));
+  removexattr(xattr, "user.mime_type");
+}
+```
+The buffer overflow happens in the kernel's `fs/ext4/xattr.c:ext4_xattr_set_entry() 
+` function, because of a [missed check](https://elixir.bootlin.com/linux/v4.15/source/fs/ext4/xattr.c#L1671) on the `size` parameter of the [`memmove()`](http://man7.org/linux/man-pages/man3/memmove.3.html) function, which is negative when trying to remove an extended attribute (`user.mime_type`) for the file `foo/bar/xattr` in the crafted image. 
+The fix is fairly straightforward in this case--adding a [value check](https://bugzilla.kernel.org/attachment.cgi?id=276147&action=diff).
+
+
+
+### Nabla and GVisor
```

Comment:
I think this summary should include Kata.  maybe the heading should be something like "Which secure containers are vulnerable?"
Then I would start the section with a sentence like "As shown, Kata containers are vulnerable, falling victim to a protocol that simply passes through too much of the filesystem-related system calls (9p).  Nabla's philosophy..."
I think the point we want to make about gVisor is that, unless they have a principled method of identifying which parts of the kernel are safe to expose (which we are not aware of), they got lucky.  Similarly, Nabla is also "lucky" if you don't believe our metric is the whole story (which it isn't) but we want to encourage the community to think about better ways to figure out what's safe to expose.